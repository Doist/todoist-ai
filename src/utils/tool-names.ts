/**
 * Centralized tool names module
 *
 * This module provides a single source of truth for all tool names used throughout the codebase.
 * Each tool should import its own name from this module to avoid hardcoded strings.
 * This prevents outdated references when tool names change.
 */

export const TOOL_NAMES = {
    // Task management tools
    TASKS_ADD_MULTIPLE: 'tasks-add-multiple',
    TASKS_COMPLETE_MULTIPLE: 'tasks-complete-multiple',
    TASKS_UPDATE_MULTIPLE: 'tasks-update-multiple',
    TASKS_SEARCH: 'tasks-search',
    TASKS_LIST_BY_DATE: 'tasks-list-by-date',
    TASKS_LIST_COMPLETED: 'tasks-list-completed',
    TASKS_LIST_FOR_CONTAINER: 'tasks-list-for-container',

    // Project management tools
    PROJECTS_LIST: 'projects-list',
    PROJECTS_MANAGE: 'projects-manage',

    // Section management tools
    SECTIONS_SEARCH: 'sections-search',
    SECTIONS_MANAGE: 'sections-manage',

    // General tools
    OVERVIEW: 'overview',
    DELETE_ONE: 'delete-one',
} as const

// Type for all tool names
export type ToolName = (typeof TOOL_NAMES)[keyof typeof TOOL_NAMES]
