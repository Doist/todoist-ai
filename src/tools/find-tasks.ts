import { GetTasksArgs, TodoistApi } from '@doist/todoist-api-typescript'
import { z } from 'zod'
import { getToolOutput } from '../mcp-helpers.js'
import type { TodoistTool } from '../todoist-tool.js'
import { getTasksByFilter, mapTask } from '../tool-helpers.js'
import { ApiLimits } from '../utils/constants.js'
import {
    generateTaskNextSteps,
    getDateString,
    previewTasks,
    summarizeList,
} from '../utils/response-builders.js'
import { MappedTask } from '../utils/test-helpers.js'
import { ToolNames } from '../utils/tool-names.js'

const { FIND_COMPLETED_TASKS, ADD_TASKS } = ToolNames

const ArgsSchema = {
    searchText: z.string().optional().describe('The text to search for in tasks.'),
    projectId: z.string().optional().describe('Find tasks in this project.'),
    sectionId: z.string().optional().describe('Find tasks in this section.'),
    parentId: z.string().optional().describe('Find subtasks of this parent task.'),
    label: z.string().optional().describe('Find tasks with this label.'),
    limit: z
        .number()
        .int()
        .min(1)
        .max(ApiLimits.TASKS_MAX)
        .default(ApiLimits.TASKS_DEFAULT)
        .describe('The maximum number of tasks to return.'),
    cursor: z
        .string()
        .optional()
        .describe(
            'The cursor to get the next page of tasks (cursor is obtained from the previous call to this tool, with the same parameters).',
        ),
}

type Args = z.infer<z.ZodObject<typeof ArgsSchema>>

const findTasks = {
    name: ToolNames.FIND_TASKS,
    description:
        'Find tasks by text search, or by project/section/parent container. At least one filter must be provided.',
    parameters: ArgsSchema,
    async execute(args, client) {
        const { searchText, projectId, sectionId, parentId, label } = args

        // Validate at least one filter is provided
        if (!searchText && !projectId && !sectionId && !parentId && !label) {
            throw new Error(
                'At least one filter must be provided: searchText, projectId, sectionId, parentId, or label',
            )
        }

        const isContainerSearch = Boolean(
            projectId || sectionId || parentId || (label && !searchText),
        )

        const { tasks, nextCursor } = isContainerSearch
            ? // If using container-based filtering or label, use direct API
              await findTasksByAttributeFilters(args, client)
            : // Text-only search using filter query
              await getTasksByFilter({
                  client,
                  query: label ? `search: ${searchText} & @${label}` : `search: ${searchText}`,
                  cursor: args.cursor,
                  limit: args.limit,
              })

        const textContent = generateTextContent({
            tasks,
            args,
            nextCursor,
            isContainerSearch,
        })

        return getToolOutput({
            textContent,
            structuredContent: {
                tasks,
                nextCursor,
                totalCount: tasks.length,
                hasMore: Boolean(nextCursor),
                appliedFilters: args,
            },
        })
    },
} satisfies TodoistTool<typeof ArgsSchema>

async function findTasksByAttributeFilters(
    { projectId, sectionId, parentId, label, limit, cursor, searchText }: Args,
    client: TodoistApi,
) {
    const taskParams: GetTasksArgs = {
        limit,
        cursor: cursor ?? null,
    }

    if (projectId) taskParams.projectId = projectId
    if (sectionId) taskParams.sectionId = sectionId
    if (parentId) taskParams.parentId = parentId
    if (label) taskParams.label = label

    const { results, nextCursor } = await client.getTasks(taskParams)
    const mappedTasks = results.map(mapTask)

    // If also has searchText, filter the results
    const tasks = searchText
        ? mappedTasks.filter(
              (task) =>
                  task.content.toLowerCase().includes(searchText.toLowerCase()) ||
                  task.description?.toLowerCase().includes(searchText.toLowerCase()),
          )
        : mappedTasks
    return { tasks, nextCursor }
}

function getContainerZeroReasonHints(args: z.infer<z.ZodObject<typeof ArgsSchema>>): string[] {
    if (args.projectId) {
        const hints = [
            args.searchText ? 'No tasks in project match search' : 'Project has no tasks yet',
        ]
        if (!args.searchText) {
            hints.push(`Use ${ADD_TASKS} to create tasks`)
        }
        return hints
    }

    if (args.sectionId) {
        const hints = [args.searchText ? 'No tasks in section match search' : 'Section is empty']
        if (!args.searchText) {
            hints.push('Tasks may be in other sections of the project')
        }
        return hints
    }

    if (args.parentId) {
        const hints = [args.searchText ? 'No subtasks match search' : 'No subtasks created yet']
        if (!args.searchText) {
            hints.push(`Use ${ADD_TASKS} with parentId to add subtasks`)
        }
        return hints
    }

    if (args.label) {
        const hints = [args.searchText ? 'No tasks match search' : 'No tasks with label were found']
        return hints
    }

    return []
}

function generateTextContent({
    tasks,
    args,
    nextCursor,
    isContainerSearch,
}: {
    tasks: MappedTask[]
    args: z.infer<z.ZodObject<typeof ArgsSchema>>
    nextCursor: string | null
    isContainerSearch: boolean
}) {
    // Generate subject and filter descriptions based on search type
    let subject: string
    const filterHints: string[] = []
    const zeroReasonHints: string[] = []

    if (isContainerSearch) {
        // Container-based search
        if (args.projectId) {
            subject = 'Tasks in project'
            filterHints.push(`in project ${args.projectId}`)
        } else if (args.sectionId) {
            subject = 'Tasks in section'
            filterHints.push(`in section ${args.sectionId}`)
        } else if (args.parentId) {
            subject = 'Subtasks'
            filterHints.push(`subtasks of ${args.parentId}`)
        } else {
            subject = 'Tasks' // fallback, though this shouldn't happen
        }

        if (args.label) {
            subject += ` with label "@${args.label}"`
            filterHints.push(`with label "@${args.label}"`)
        }

        // Add search text filter if present
        if (args.searchText) {
            subject += ` matching "${args.searchText}"`
            filterHints.push(`containing "${args.searchText}"`)
        }

        // Container-specific zero result hints
        if (tasks.length === 0) {
            zeroReasonHints.push(...getContainerZeroReasonHints(args))
        }
    } else {
        // Text-only search
        subject = `Search results for "${args.searchText}"`
        filterHints.push(`matching "${args.searchText}"`)

        if (args.label) {
            subject += ` filtered by label "@${args.label}"`
            filterHints.push(`filtered by label "@${args.label}"`)
        }

        if (tasks.length === 0) {
            zeroReasonHints.push('Try broader search terms')
            zeroReasonHints.push(`Check completed tasks with ${FIND_COMPLETED_TASKS}`)
            zeroReasonHints.push('Verify spelling and try partial words')
        }
    }

    // Generate contextual next steps
    const now = new Date()
    const todayDateString = getDateString(now)
    const nextSteps = generateTaskNextSteps('listed', tasks, {
        hasToday: tasks.some((task) => task.dueDate === todayDateString),
        hasOverdue: tasks.some((task) => task.dueDate && new Date(task.dueDate) < now),
    })

    return summarizeList({
        subject,
        count: tasks.length,
        limit: args.limit,
        nextCursor: nextCursor ?? undefined,
        filterHints,
        previewLines: previewTasks(tasks),
        zeroReasonHints,
        nextSteps,
    })
}

export { findTasks }
