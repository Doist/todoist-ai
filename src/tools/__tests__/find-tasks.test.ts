import type { TodoistApi } from '@doist/todoist-api-typescript'
import { jest } from '@jest/globals'
import { getTasksByFilter } from '../../tool-helpers.js'
import {
    type MappedTask,
    TEST_ERRORS,
    TEST_IDS,
    TODAY,
    createMappedTask,
    createMockApiResponse,
    createMockTask,
    extractStructuredContent,
    extractTextContent,
} from '../../utils/test-helpers.js'
import { ToolNames } from '../../utils/tool-names.js'
import { findTasks } from '../find-tasks.js'

jest.mock('../../tool-helpers', () => {
    const actual = jest.requireActual('../../tool-helpers') as typeof import('../../tool-helpers')
    return {
        getTasksByFilter: jest.fn(),
        mapTask: actual.mapTask,
    }
})

const { FIND_TASKS, UPDATE_TASKS, FIND_COMPLETED_TASKS } = ToolNames

const mockGetTasksByFilter = getTasksByFilter as jest.MockedFunction<typeof getTasksByFilter>

// Mock the Todoist API
const mockTodoistApi = {
    getTasks: jest.fn(),
} as unknown as jest.Mocked<TodoistApi>

describe(`${FIND_TASKS} tool`, () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    describe('searching tasks', () => {
        it('should search tasks and return results', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Task containing search term',
                    description: 'Description with more details',
                    labels: ['work'],
                }),
                createMappedTask({
                    id: TEST_IDS.TASK_2,
                    content: 'Another matching task',
                    priority: 2,
                    sectionId: TEST_IDS.SECTION_1,
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: 'cursor-for-next-page' }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasks.execute(
                {
                    searchText: 'important meeting',
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or',
                },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: 'search: important meeting',
                cursor: undefined,
                limit: 10,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    tasks: expect.any(Array),
                    totalCount: 2,
                    hasMore: true,
                    nextCursor: 'cursor-for-next-page',
                    appliedFilters: {
                        searchText: 'important meeting',
                        limit: 10,
                        labels: [],
                        labelsOperator: 'or',
                    },
                }),
            )
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    tasks: expect.any(Array),
                }),
            )
        })

        it.each([
            {
                name: 'custom limit',
                params: {
                    searchText: 'project update',
                    limit: 5,
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectedQuery: 'search: project update',
                expectedLimit: 5,
                expectedCursor: undefined,
            },
            {
                name: 'pagination cursor',
                params: {
                    searchText: 'follow up',
                    limit: 20,
                    cursor: 'cursor-from-first-page',
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectedQuery: 'search: follow up',
                expectedLimit: 20,
                expectedCursor: 'cursor-from-first-page',
            },
        ])(
            'should handle $name',
            async ({ params, expectedQuery, expectedLimit, expectedCursor }) => {
                const mockTask = createMappedTask({ content: 'Test result' })
                const mockResponse = { tasks: [mockTask], nextCursor: null }
                mockGetTasksByFilter.mockResolvedValue(mockResponse)

                const result = await findTasks.execute(params, mockTodoistApi)

                expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                    client: mockTodoistApi,
                    query: expectedQuery,
                    cursor: expectedCursor,
                    limit: expectedLimit,
                })
                // Verify result is a concise summary
                expect(extractTextContent(result)).toMatchSnapshot()

                // Verify structured content
                const structuredContent = extractStructuredContent(result)
                expect(structuredContent).toEqual(
                    expect.objectContaining({
                        tasks: expect.any(Array),
                        totalCount: 1,
                        hasMore: false,
                        nextCursor: null,
                        appliedFilters: expect.objectContaining({
                            searchText: params.searchText,
                            limit: expectedLimit,
                        }),
                    }),
                )
                expect(structuredContent).toEqual(
                    expect.objectContaining({
                        tasks: expect.any(Array),
                    }),
                )
            },
        )

        it.each([
            { searchText: '@work #urgent "exact phrase"', description: 'special characters' },
            { searchText: 'nonexistent keyword', description: 'empty results' },
        ])('should handle search with $description', async ({ searchText }) => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasks.execute(
                { searchText, limit: 10, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: `search: ${searchText}`,
                cursor: undefined,
                limit: 10,
            })
            // Verify result is a concise summary
            expect(extractTextContent(result)).toMatchSnapshot()

            // Verify structured content for empty results
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    tasks: [],
                    totalCount: 0,
                    hasMore: false,
                    nextCursor: null,
                    appliedFilters: expect.objectContaining({
                        searchText: searchText,
                    }),
                }),
            )
        })
    })

    describe('validation', () => {
        it('should require at least one filter parameter', async () => {
            await expect(
                findTasks.execute({ limit: 10, labels: [], labelsOperator: 'or' }, mockTodoistApi),
            ).rejects.toThrow(
                'At least one filter must be provided: searchText, projectId, sectionId, or parentId',
            )
        })
    })

    describe('container filtering', () => {
        it.each([
            {
                name: 'project',
                params: {
                    projectId: TEST_IDS.PROJECT_TEST,
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectedApiParam: { projectId: TEST_IDS.PROJECT_TEST },
                tasks: [createMockTask({ content: 'Project task' })],
            },
            {
                name: 'section',
                params: {
                    sectionId: TEST_IDS.SECTION_1,
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectedApiParam: { sectionId: TEST_IDS.SECTION_1 },
                tasks: [createMockTask({ content: 'Section task' })],
            },
            {
                name: 'parent task',
                params: {
                    parentId: TEST_IDS.TASK_1,
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectedApiParam: { parentId: TEST_IDS.TASK_1 },
                tasks: [createMockTask({ content: 'Subtask' })],
            },
        ])('should find tasks in $name', async ({ params, expectedApiParam, tasks }) => {
            mockTodoistApi.getTasks.mockResolvedValue(createMockApiResponse(tasks))

            const result = await findTasks.execute(params, mockTodoistApi)

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 10,
                cursor: null,
                ...expectedApiParam,
            })

            expect(extractTextContent(result)).toMatchSnapshot()

            const structuredContent = extractStructuredContent(result)
            expect(structuredContent).toEqual(
                expect.objectContaining({
                    tasks: expect.any(Array),
                    totalCount: tasks.length,
                    hasMore: false,
                    appliedFilters: params,
                }),
            )
        })

        it('should handle combined search text and container filtering', async () => {
            const tasks = [
                createMockTask({
                    id: '8485093749',
                    content: 'relevant task',
                    description: 'contains search term',
                }),
                createMockTask({
                    id: '8485093750',
                    content: 'other task',
                    description: 'different content',
                }),
            ]
            mockTodoistApi.getTasks.mockResolvedValue(createMockApiResponse(tasks))

            const result = await findTasks.execute(
                {
                    projectId: TEST_IDS.PROJECT_TEST,
                    searchText: 'relevant',
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or',
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 10,
                cursor: null,
                projectId: TEST_IDS.PROJECT_TEST,
            })

            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.tasks).toHaveLength(1)
            expect(structuredContent.tasks).toEqual([
                expect.objectContaining({ content: 'relevant task' }),
            ])
        })

        it('should handle empty containers', async () => {
            mockTodoistApi.getTasks.mockResolvedValue(createMockApiResponse([]))

            const result = await findTasks.execute(
                {
                    sectionId: 'empty-section',
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or',
                },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toContain('Section is empty')
            expect(textContent).toContain('Tasks may be in other sections of the project')
        })

        it('should handle pagination with containers', async () => {
            mockTodoistApi.getTasks.mockResolvedValue({
                results: [],
                nextCursor: 'next-cursor',
            })

            const result = await findTasks.execute(
                {
                    projectId: TEST_IDS.PROJECT_TEST,
                    limit: 25,
                    cursor: 'current-cursor',
                    labels: [],
                    labelsOperator: 'or',
                },
                mockTodoistApi,
            )

            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 25,
                cursor: 'current-cursor',
                projectId: TEST_IDS.PROJECT_TEST,
            })

            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.hasMore).toBe(true)
            expect(structuredContent.nextCursor).toBe('next-cursor')
        })
    })

    describe('container error handling', () => {
        it('should propagate API errors for container queries', async () => {
            const apiError = new Error('API Error: Project not found')
            mockTodoistApi.getTasks.mockRejectedValue(apiError)

            await expect(
                findTasks.execute(
                    { projectId: 'non-existent', limit: 10, labels: [], labelsOperator: 'or' },
                    mockTodoistApi,
                ),
            ).rejects.toThrow('API Error: Project not found')
        })
    })

    describe('next steps logic', () => {
        it('should suggest different actions when hasOverdue is true', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Overdue search result',
                    dueDate: '2025-08-10', // Past date
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasks.execute(
                { searchText: 'overdue tasks', limit: 10, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            // Should prioritize overdue tasks in next steps
            expect(textContent).toContain(`Use ${UPDATE_TASKS} to modify priorities or due dates`)
        })

        it('should suggest today tasks when hasToday is true', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Task due today',
                    dueDate: TODAY,
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasks.execute(
                { searchText: 'today tasks', limit: 10, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            // Should suggest today-focused actions
            expect(textContent).toContain(`Use ${UPDATE_TASKS} to modify priorities or due dates`)
        })

        it('should provide different next steps for regular tasks', async () => {
            const mockTasks = [
                createMappedTask({
                    id: TEST_IDS.TASK_1,
                    content: 'Regular future task',
                    dueDate: '2025-08-25', // Future date
                }),
            ]
            const mockResponse = { tasks: mockTasks, nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasks.execute(
                { searchText: 'future tasks', limit: 10, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain(`Use ${UPDATE_TASKS} to modify priorities or due dates`)
        })

        it('should provide helpful suggestions for empty search results', async () => {
            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            const result = await findTasks.execute(
                { searchText: 'nonexistent', limit: 10, labels: [], labelsOperator: 'or' },
                mockTodoistApi,
            )

            const textContent = extractTextContent(result)
            expect(textContent).toMatchSnapshot()
            expect(textContent).toContain('Try broader search terms')
            expect(textContent).toContain(`Check completed tasks with ${FIND_COMPLETED_TASKS}`)
            expect(textContent).toContain('Verify spelling and try partial words')
        })
    })

    describe('label filtering', () => {
        it.each([
            {
                name: 'text search with single label OR operator',
                params: {
                    searchText: 'important meeting',
                    limit: 10,
                    labels: ['@work'],
                    labelsOperator: 'or' as const,
                },
                expectedQuery: 'search: important meeting & (@work)',
            },
            {
                name: 'text search with multiple labels AND operator',
                params: {
                    searchText: 'project update',
                    limit: 15,
                    labels: ['@work', '@urgent'],
                    labelsOperator: 'and' as const,
                },
                expectedQuery: 'search: project update & (@work  &  @urgent)',
            },
            {
                name: 'text search with multiple labels OR operator',
                params: {
                    searchText: 'follow up',
                    limit: 20,
                    labels: ['@personal', '@shopping'],
                    labelsOperator: 'or' as const,
                },
                expectedQuery: 'search: follow up & (@personal  |  @shopping)',
            },
        ])(
            'should filter tasks by labels in text search: $name',
            async ({ params, expectedQuery }) => {
                const mockTasks = [
                    createMappedTask({
                        id: TEST_IDS.TASK_1,
                        content: 'Task with work label',
                        labels: ['work'],
                    }),
                ]
                const mockResponse = { tasks: mockTasks, nextCursor: null }
                mockGetTasksByFilter.mockResolvedValue(mockResponse)

                const result = await findTasks.execute(params, mockTodoistApi)

                expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                    client: mockTodoistApi,
                    query: expectedQuery,
                    cursor: undefined,
                    limit: params.limit,
                })

                const structuredContent = extractStructuredContent(result)
                expect(structuredContent.appliedFilters).toEqual(
                    expect.objectContaining({
                        searchText: params.searchText,
                        labels: params.labels,
                        labelsOperator: params.labelsOperator,
                    }),
                )
            },
        )

        it.each([
            {
                name: 'project filter with labels',
                params: {
                    projectId: TEST_IDS.PROJECT_TEST,
                    limit: 10,
                    labels: ['@important'],
                    labelsOperator: 'or' as const,
                },
                expectedApiParam: { projectId: TEST_IDS.PROJECT_TEST },
            },
            {
                name: 'section filter with multiple labels',
                params: {
                    sectionId: TEST_IDS.SECTION_1,
                    limit: 10,
                    labels: ['@work', '@urgent'],
                    labelsOperator: 'and' as const,
                },
                expectedApiParam: { sectionId: TEST_IDS.SECTION_1 },
            },
            {
                name: 'parent task filter with labels',
                params: {
                    parentId: TEST_IDS.TASK_1,
                    limit: 10,
                    labels: ['@personal'],
                    labelsOperator: 'or' as const,
                },
                expectedApiParam: { parentId: TEST_IDS.TASK_1 },
            },
        ])(
            'should apply label filtering to container searches: $name',
            async ({ params, expectedApiParam }) => {
                const allTasks = [
                    createMockTask({
                        id: '1',
                        content: 'Task with matching label',
                        labels: params.labels,
                    }),
                    createMockTask({
                        id: '2',
                        content: 'Task without matching label',
                        labels: ['other'],
                    }),
                ]
                mockTodoistApi.getTasks.mockResolvedValue(createMockApiResponse(allTasks))

                const result = await findTasks.execute(params, mockTodoistApi)

                expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                    limit: 10,
                    cursor: null,
                    ...expectedApiParam,
                })

                // Should filter results client-side based on labels
                const structuredContent = extractStructuredContent(result)
                if (params.labelsOperator === 'and') {
                    // AND operation: task must have all specified labels
                    expect(structuredContent.tasks).toEqual(
                        expect.arrayContaining([
                            expect.objectContaining({
                                labels: expect.arrayContaining(params.labels),
                            }),
                        ]),
                    )
                } else {
                    // OR operation: task must have at least one of the specified labels
                    expect((structuredContent.tasks as MappedTask[]).length).toBeGreaterThanOrEqual(
                        0,
                    )
                }
            },
        )

        it('should handle empty labels array', async () => {
            const params = {
                searchText: 'test',
                limit: 10,
                labels: [],
                labelsOperator: 'or' as const,
            }

            const mockResponse = { tasks: [], nextCursor: null }
            mockGetTasksByFilter.mockResolvedValue(mockResponse)

            await findTasks.execute(params, mockTodoistApi)

            expect(mockGetTasksByFilter).toHaveBeenCalledWith({
                client: mockTodoistApi,
                query: 'search: test',
                cursor: undefined,
                limit: 10,
            })
        })

        it('should combine search text, container, and label filters', async () => {
            const params = {
                projectId: TEST_IDS.PROJECT_TEST,
                searchText: 'important',
                limit: 10,
                labels: ['urgent'],
                labelsOperator: 'or' as const,
            }

            const allTasks = [
                createMockTask({
                    id: '1',
                    content: 'important task',
                    description: 'urgent work',
                    labels: ['urgent'],
                }),
                createMockTask({
                    id: '2',
                    content: 'other task',
                    description: 'not important',
                    labels: ['work'],
                }),
            ]
            mockTodoistApi.getTasks.mockResolvedValue(createMockApiResponse(allTasks))

            const result = await findTasks.execute(params, mockTodoistApi)

            // Should call API with container filter
            expect(mockTodoistApi.getTasks).toHaveBeenCalledWith({
                limit: 10,
                cursor: null,
                projectId: TEST_IDS.PROJECT_TEST,
            })

            // Should filter results by search text AND labels
            const structuredContent = extractStructuredContent(result)
            expect(structuredContent.tasks).toEqual([
                expect.objectContaining({
                    content: 'important task',
                    labels: expect.arrayContaining(['urgent']),
                }),
            ])
        })
    })

    describe('error handling', () => {
        it.each([
            {
                error: 'At least one filter must be provided: searchText, projectId, sectionId, or parentId',
                params: { limit: 10, labels: [], labelsOperator: 'or' as const },
                expectValidation: true,
            },
            {
                error: TEST_ERRORS.API_RATE_LIMIT,
                params: {
                    searchText: 'any search term',
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectValidation: false,
            },
            {
                error: TEST_ERRORS.INVALID_CURSOR,
                params: {
                    searchText: 'test',
                    cursor: 'invalid-cursor-format',
                    limit: 10,
                    labels: [],
                    labelsOperator: 'or' as const,
                },
                expectValidation: false,
            },
        ])('should propagate $error', async ({ error, params, expectValidation }) => {
            if (!expectValidation) {
                mockGetTasksByFilter.mockRejectedValue(new Error(error))
            }
            await expect(findTasks.execute(params, mockTodoistApi)).rejects.toThrow(error)
        })
    })
})
