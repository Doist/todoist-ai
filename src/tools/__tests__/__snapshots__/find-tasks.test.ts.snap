// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`find-tasks tool container filtering should find tasks in parent task 1`] = `
"Subtasks: 1 (limit 10).
Filter: subtasks of 8485093748.
Preview:
    Subtask • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks"
`;

exports[`find-tasks tool container filtering should find tasks in project 1`] = `
"Tasks in project: 1 (limit 10).
Filter: in project 6cfCcrrCFg2xP94Q.
Preview:
    Project task • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks"
`;

exports[`find-tasks tool container filtering should find tasks in section 1`] = `
"Tasks in section: 1 (limit 10).
Filter: in section section-123.
Preview:
    Section task • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks"
`;

exports[`find-tasks tool next steps logic should provide different next steps for regular tasks 1`] = `
"Search results for "future tasks": 1 (limit 10).
Filter: matching "future tasks".
Preview:
    Regular future task • due 2025-08-25 • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks tool next steps logic should provide helpful suggestions for empty search results 1`] = `
"Search results for "nonexistent": 0 (limit 10).
Filter: matching "nonexistent".
No results. Try broader search terms; Check completed tasks with find-completed-tasks; Verify spelling and try partial words."
`;

exports[`find-tasks tool next steps logic should suggest different actions when hasOverdue is true 1`] = `
"Search results for "overdue tasks": 1 (limit 10).
Filter: matching "overdue tasks".
Preview:
    Overdue search result • due 2025-08-10 • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks tool next steps logic should suggest today tasks when hasToday is true 1`] = `
"Search results for "today tasks": 1 (limit 10).
Filter: matching "today tasks".
Preview:
    Task due today • due 2025-08-17 • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Focus on overdue items first to get back on track"
`;

exports[`find-tasks tool searching tasks should handle custom limit 1`] = `
"Search results for "project update": 1 (limit 5).
Filter: matching "project update".
Preview:
    Test result • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks"
`;

exports[`find-tasks tool searching tasks should handle pagination cursor 1`] = `
"Search results for "follow up": 1 (limit 20).
Filter: matching "follow up".
Preview:
    Test result • P1 • id=8485093748
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks"
`;

exports[`find-tasks tool searching tasks should handle search with empty results 1`] = `
"Search results for "nonexistent keyword": 0 (limit 10).
Filter: matching "nonexistent keyword".
No results. Try broader search terms; Check completed tasks with find-completed-tasks; Verify spelling and try partial words."
`;

exports[`find-tasks tool searching tasks should handle search with special characters 1`] = `
"Search results for "@work #urgent "exact phrase"": 0 (limit 10).
Filter: matching "@work #urgent "exact phrase"".
No results. Try broader search terms; Check completed tasks with find-completed-tasks; Verify spelling and try partial words."
`;

exports[`find-tasks tool searching tasks should search tasks and return results 1`] = `
"Search results for "important meeting": 2 (limit 10), more available.
Filter: matching "important meeting".
Preview:
    Task containing search term • P1 • id=8485093748
    Another matching task • P2 • id=8485093749
Next:
- Use update-tasks to modify priorities or due dates
- Use complete-tasks to mark finished tasks
- Pass cursor 'cursor-for-next-page' to fetch more results."
`;
